### ISSUE TITLE
Support emitting configuration feedback information from kpt fns

### ISSUE THREAD

**Issue by dflemstr** [GoogleContainerTools/kpt#291]
We would like to offer structured feedback to users when linting/validating KRM objects that are getting transformed as part of a `kpt fn` pipeline.  This feedback would originate from existing `kpt` functions such as [opa-gatekeeper](https://github.com/GoogleContainerTools/kpt-functions-sdk/blob/a4455d83d68d573f687a25211f3b108d8179754b/go/pkg/functions/gatekeeper/validate.go), emerging `kpt` functions that use `kubeval`, and other functions using proprietary linting rules.

Some interesting constraints:

  * It would be nice if at the end of the pipeline run, we end up with all of the feedback from all of the fns, so we should not have fns exit when they encounter errors.
  * We definitely need to support "hard" and "soft" errors (i.e. warnings/notes).
  * We will want to build quite intelligent linters, so it would be nice to offer support for highlighting the path to the line/column/offending manifest key, offer a suggestion for what it should be instead, refer to context (the typical `note: caused by this other related field having value X` e.g. `note: caused by setting region to global`).  What we will want to do at Spotify is to hook up this feedback with the [GitHub checks API](https://developer.github.com/v3/checks/), so it would be nice to see how much ground we could cover to help integrate with that.

Suggestions for implementations:

1. Support a custom KRM resource that encapsulates feedback, sketch:

```yaml
apiVersion: kpt.something/vsomething
kind: Feedback
spec:
  items:
    - kind: error
      resourceRef:
        apiVersion: apps/v1
        kind: Deployment
        metadata:
          name: foobar
      message: "Value exceeds the namespace quota, reduce the value to make the pod schedulable"
      field: "spec.template.spec.containers[3].resources.limits.cpu"
      currentValue: "200"
      suggestedValue: "2"
    - kind: note
      # ... similar to the above example
    - kind: warning
      # ... similar to the above example
```

2. Support something similar to the above, but by appending information to the affected object directly:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: foobar
spec:
  # ...
feedback:
  - kind: error
  # ... similar to the above example
```

Let me know what direction you would prefer and we could start looking at implementing this

**Comment by pwittrock** [comment]
cc @frankfarzan @prachirp

@dflemstr 

Thanks for the issue!  This makes a lot of sense.  How are you running the functions -- e.g. is it `kpt fn run DIR/`, `kpt fn run DIR/ --image gcr.io/`, through a CICD pipeline directly wiring the pipeline together, etc?

I like the first option in your proposal.  We could add this to the function input/output spec as an optional piece of data emitted by functions as part of the `ResourceList` (e.g. along side the `functionConfig` field).

Roughly, that would look something like:

- Parse it here: https://github.com/kubernetes-sigs/kustomize/blob/master/kyaml/kio/byteio_reader.go#L150
- Save it here: https://github.com/kubernetes-sigs/kustomize/blob/master/kyaml/kio/filters/container.go#L284
- Aggregate it here: https://github.com/kubernetes-sigs/kustomize/blob/master/kyaml/runfn/runfn.go#L150

A few additional possibilities for consideration:

> It would be nice if at the end of the pipeline run, we end up with all of the feedback from all of the fns, so we should not have fns exit when they encounter errors.

We could get this from the structured output, additionally we could have the function annotation on the KRM resource triggering the function explicitly specify this behavior on error -- I could image `stop`, `continue` etc

e.g. something along the lines of:

    ```yaml
    metadata:
      annotations:
        config.kubernetes.io/function: |
          continueOnError: true
          container:
            image: gcr.io/example.com/image:version
    ```

I can see us wanting both knobs -- if a validation function doesn't implement writing back the results (e.g. such as the validators you use today), you probably want a knob to explicitly define how the pipeline is orchestrated.

> We definitely need to support "hard" and "soft" errors (i.e. warnings/notes).

This makes sense.  The structured output approach you listed makes sense for this.  We could additionally support the use exit codes as a lighter weight approach for function authors.  The latter may be a nice to have.

> It would be nice if at the end of the pipeline run, we end up with all of the feedback from all of the fns, so we should not have fns exit when they encounter errors.

We can 1) aggregate this for the fns that emit structured output.  2) for the fns that don't emit structured output, but are configured to not terminate the pipeline on error, we can use the stderr and exist code to construct structured output that could be aggregated.

> We will want to build quite intelligent linters, so it would be nice to offer support for highlighting the path to the line/column/offending manifest key, offer a suggestion for what it should be instead, refer to context (the typical note: caused by this other related field having value X e.g. note: caused by setting region to global). What we will want to do at Spotify is to hook up this feedback with the GitHub checks API, so it would be nice to see how much ground we could cover to help integrate with that.

This makes sense.  The linting functions themselves would need to support providing the structured output.  We might even want that, given an error for a particular resource/field, the pipeline orchestrator could also provide the filename and line number (the go-yaml libraries should give this to use).  Note this would only work if the pipeline didn't transform the input resources (the line numbers wouldn't be updated to reflect the changes).

@frankfarzan @dflemstr: WDYT of the next steps (very roughly):

Short term wins:

- Add support for not terminating the pipeline on validation errors, but failing at the end -- using annotations on the triggering functions
  - This only makes sense if you are using `kpt fn run` to run the function pipelines -- otherwise this isn't helpful
  - This is a relatively small to addition but would unblock you in the short term and without requiring rewriting the functions themselves

Long term goal:

- Propose the output metadata type definition -- this can be OpenAPI, a go struct, an example, etc
  - Come up with the right name for this field -- maybe `FunctionResults`? (IDK)
- Define how `FunctionResults` would be aggregated across multiple function invocations
- Wire into the kpt function orchestrator (e.g. `kpt fn run`) support for parsing and aggregating the metadata type
  - Read the `FunctionResults` from the `ByteReader` and store to a local field -- similar to what is done for `FuctionConfig`
  - Save the `FunctionResults` from the `ContainerFilter` so they are accessible after the pipeline terminates.
  - `RunFns` aggregates the `FunctionConfig` results
- ContainerFilter + RunFns supports hard / soft failures + continue-on-error based of output metadata
- Write the aggregated output to some location -- may be defined by a flag to `kpt fn run`

**Comment by pwittrock** [comment]
@frankfarzan Is OOO right now, but should definitely be included in this discussion

**Comment by dflemstr** [comment]
Your next steps sound great, we have added this to our backlog and will need to get back to it in a little while since we have some other priorities to get through first.  Being able to get back to the information in this PR when we're ready will be immensely useful, of course.

Right now, in our CI/CD prototype, we are running a fixed list of functions with `kpt fn source ... | kpt fn run ... | kpt fn run ... | kpt fn sink`.  Of course we can tweak that to be something more idiomatic, but for now we don't expect our users to have `Kptfile`s in their repositories, since we will only use fns for validation and not transformation for now.

**Comment by pwittrock** [comment]
Ah, that makes sense now.  The errors would need to be propagated through the output then so they can be aggregated in the sink without failing early.

We may then want to introduce a flag `kpt fn run --fail-on-error=false` and then make sure the FunctionResults were propagated between invocations.

**Comment by frankfarzan** [comment]
I agree with the proposal as an optional feature with the caveat that for the majority of functions implemented, it’s sufficient to use exit codes and stderr logs.

One thing to add to the discussion: We need to delineate responsibilities between the function implementation and the orchestrator. In the future (See FR #226), we want to enable running the pipeline using different orchestrators (not just using `kpt run` locally).

I see these 3 different responsibilities: 

1. Functions: Optionally emit a structured feedback i.e. `FunctionResults`. In terms of implementation, agree with Phill that option 1 (using `ResourceList`) is preferred to appending to resources (Which hacks the schema).
2. Orchestrator: Determines whether a pipeline should be short circuited depending on the exit code + `FunctionResults`. There are features in Tekton, Argo Workflows, and other orchestrators that this maps to (Usually called “conditionals”).
3. Reporter: Knows how to aggregate a list of `FunctionResults` resources and do something useful with it. It’d be interesting if we could model this as a Sink function, where the input is the list of Feedback resources, and the function calls GitHub API for example.

It’s going to be tricky to have source line-level feedback because:
- Functions may transform configs as Phill pointed out
- Roundtripping to native data types doesn’t preserve YAML comments depending on the language. kpt (kyaml) uses YAML AST which preserves comments, but the API is too low-level. It’s on the roadmap for the orchestrator to handle preserving comments to simplify the burden on function implementers.

**Comment by dflemstr** [comment]
Another option (Which we are using right now, since the Typescript SDK uses pretty decent structured error formatting...) is to define a stricter schema for stderr output, of course.  I'm not sure how reliable that would be, as such output might need to be intermingled with regular debug output and such.

**Comment by pwittrock** [comment]
```yaml
kind: ResourceList
results:
- type: error # one of ["error", "warning", "note"]
  message: "Value exceeds the namespace quota, reduce the value to make the pod schedulable"
  annotations: # arbitrary metadata about the result
    error: "field-error"
  resource:
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      name: foobar
  file:
    path: deploy.yaml # read from annotation if present -- needed for kustomize patch cases
    index: 0 # read from annotation if present -- needed for kustomize patch cases
    line: 40 # yaml libraries can give this to us -- may not be correct if mutated
  field:
    path: "spec.template.spec.containers[3].resources.limits.cpu"
    currentValue: "200"
    suggestedValue: "2" 
- type: warning
  ...
- type: note
   ...
```

**Comment by frankfarzan** [comment]
We also need to add this to the Spec:
https://github.com/kubernetes-sigs/kustomize/blob/master/cmd/config/docs/api-conventions/functions-spec.md

**Comment by frankfarzan** [comment]
Some questions regarding https://github.com/GoogleContainerTools/kpt/issues/291#issuecomment-608687816:

- What does ObjectError kind conceptually represent? Any reason, we can simplify this to:

```yaml
kind: ResourceList
items: [...] # already part of io spec
functionConfig: {...} # already part of io spec
results:  # aggregate results 
  - severity: error # one of ["error", "warning", "note"]
    tags: # arbitrary metadata about the result
      error-type: "field"
    message: "Value exceeds the namespace quota, reduce the value to make the pod schedulable"
    c: # key to lookup the resourtace
      apiVersion: apps/v1
      kind: Deployment
      name: foo
      namespace: bar
    file:
      # optional if present as annotation
      path: deploy.yaml # read from annotation if present -- needed for kustomize patch cases
      # optional if present as annotation
      index: 0 # read from annotation if present -- needed for kustomize patch cases
    field:
      path: "spec.template.spec.containers[3].resources.limits.cpu"
      currentValue: "200"
      suggestedValue: "2" 
  - type: warning
    ...
```
- What's the type of `field.currentValue|suggestedValue`? I'm guessing we only want to support primitives (number | string | boolean) and not objects?
- "type" -> "severity"? Type is an overloaded word

**Comment by frankfarzan** [comment]
Defined the schema in TS here:

https://github.com/GoogleContainerTools/kpt-functions-sdk/pull/133

This should highlight any ambiguities in the proposed spec.

**Comment by frankfarzan** [comment]
To give this functionality a try, you can use the demo functions with `results` tag:

```
$ kpt pkg get https://github.com/GoogleContainerTools/kpt-functions-sdk.git/example-configs example-configs
$ cd example-configs
$ kpt fn source . | docker run -i gcr.io/kpt-functions/validate-rolebinding:results -d subject_name=bob@foo-corp.com
```

**Comment by pwittrock** [comment]
#kubernetes-sigs/kustomize/2359

**Comment by frankfarzan** [comment]
Added `suggest-psp` demo function:

```
$ kpt pkg get https://github.com/GoogleContainerTools/kpt-functions-sdk.git/example-configs example-configs
$ cd example-configs
$ kpt fn source . | docker run -e STRUCTURED_RESULTS=1 -i gcr.io/kpt-functions/suggest-psp:dev
```

Outputs:

```yaml
results:
- message: Suggest explicitly disabling privilege escalation
  severity: warn
  tags:
    category: security
  resourceRef:
    apiVersion: policy/v1beta1
    kind: PodSecurityPolicy
    namespace: ''
    name: psp
  file:
    path: test-data/source/foo-yaml/foo-corp-1.0.0/podsecuritypolicy_psp.yaml
  field:
    path: spec.allowPrivilegeEscalation
    suggestedValue: false
```

**Comment by pwittrock** [comment]
https://googlecontainertools.github.io/kpt/guides/producer/functions/#validation
